tests/fuzz/fuzz-codex32.c

```diff
@@ -0,0 +1,255 @@
+	memcpy(dup->id, src->id, sizeof(dup->id));
+	dup->share_idx = src->share_idx;
+	dup->payload = tal_dup_arr(dup, u8, src->payload,
+				   tal_bytelen(src->payload), 0);
+	dup->type = src->type;
+	return dup;
+}
+
+static bool codex32_fields_valid_for_encoding(const struct codex32 *parts)
+{
+	/* Check threshold */
+	if (parts->threshold > 9 || parts->threshold == 1)
+		return false;
+
+	/* Check id: must be 4 characters, each in bech32 charset and ASCII */
+	if (strlen(parts->id) != 4)
+		return false;
+
+	for (int i = 0; i < 4; i++) {
+		unsigned char c = parts->id[i];
+		if (c == 0 || c >= 128)
+			return false;
+		if (bech32_charset_rev[c] == -1)
+			return false;
+	}
+
+	/* Check HRP: must be 2 characters, no embedded zeros, and ASCII */
+	if (strlen(parts->hrp) != 2)
+		return false;
+
+	if (parts->hrp[0] == 0)
+		return false;
+
+	for (size_t i = 0; i < strlen(parts->hrp); i++) {
+		unsigned char c = parts->hrp[i];
+		if (c == 0 || c >= 128)
+			return false;
+	}
+
+	/* Check share index */
+	unsigned char si = parts->share_idx;
+	if (si == 0 || si >= 128)
+		return false;
+	if (bech32_charset_rev[si] == -1)
+		return false;
+
+	return true;
+}
+
+void init(int *argc, char ***argv)
+{
+	common_setup("fuzzer");
+}
+
+/* Custom mutator with structure-aware and byte-level mutations */
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size,
+				size_t max_size, unsigned int seed)
+{
+	srand(seed);
+	char *str = to_string(tmpctx, fuzz_data, size);
+	char *fail;
+	struct codex32 *parts = codex32_decode(tmpctx, NULL, str, &fail);
+
+	/* If valid, try structure-aware mutation */
+	if (parts) {
+		/* Mutate a random component */
+		switch(rand() % 4) {
+		case 0: /* Mutate HRP arbitrarily */
+			{
+				size_t hrp_len = strlen(parts->hrp);
+				size_t max_hrp_len = 10; /* Reasonable max */
+				char *new_hrp = tal_arr(parts, char, max_hrp_len + 1);
+				size_t new_hrp_len = LLVMFuzzerMutate((u8 *)new_hrp,
+								hrp_len, max_hrp_len);
+				new_hrp[new_hrp_len] = '\0';
+				parts->hrp = new_hrp;
+			}
+			break;
+
+		case 1: /* Mutate threshold to any value */
+			parts->threshold = rand();
+			break;
+
+		case 2: /* Mutate ID arbitrarily */
+			{
+				size_t id_len = sizeof(parts->id) - 1;
+				LLVMFuzzerMutate((u8 *)parts->id, id_len, id_len);
+				parts->id[id_len] = '\0';
+			}
+			break;
+
+		case 3: /* Mutate payload */
+			{
+				size_t old_size = tal_bytelen(parts->payload);
+				tal_resize(&parts->payload, max_size);
+				size_t new_size = LLVMFuzzerMutate((u8 *)parts->payload, old_size, max_size);
+				tal_resize(&parts->payload, new_size);
+			}
+			break;
+		}
+
+		/* Only try to re-encode if parts are valid */
+		if (codex32_fields_valid_for_encoding(parts)) {
+			char *reencoded;
+			const char *err = codex32_secret_encode(tmpctx, parts->hrp, parts->id,
+								parts->threshold, parts->payload,
+								tal_bytelen(parts->payload), &reencoded);
+			if (!err) {
+				size_t len = tal_bytelen(reencoded);
```

Comment:
I think we should drop the null byte, since `run` adds a null byte.
```suggestion
				size_t len = tal_bytelen(reencoded) - 1;
```